### PR TITLE
ENH: Automated Package Building and Deployment

### PR DESCRIPTION
**Pull request by esavary**
This PR introduces new jobs in the build_test_deploy CircleCI workflow for building and deploying the package to PyPI.
related to #152

### CODE REVIEW

**Comment by esavary** [comment]
@effigies  @oesteban , I'm wondering if there's a clever way to test the code I added in the CircleCi workflow without needing to push a tag. (To prevent misunderstanding, @oesteban , you mentioned yesterday using TestPyPI to test the package, but what I'm asking here is a bit different—I'm looking for a way to specifically test the new CircleCI routine itself, not the package.)  Any suggestions on this would be helpful!

**Comment by effigies** [comment]
For new workflows, I would strongly recommend adopting https://github.com/hynek/structlog/blob/main/.github/workflows/pypi-package.yml more or less wholesale. Its first action produces easily inspectable reports in the actions tab: https://github.com/hynek/structlog/actions/runs/8584514099

image

Since these actions are maintained, they are less likely to bitrot than custom scripts in CircleCI.

**Comment by effigies** [comment]
We can create a deployment environment that requires a maintainer to approve the release before pushing.

**Comment by effigies** [comment]
Here is an example of using an environment:

https://github.com/nipy/nibabel/blob/a2e5dee05cf374c22670ff9fd0d385ce366eb495/.github/workflows/test.yml#L199-L212

I created one here called "Release"

**Comment by esavary** [comment]
> Here is an example of using an environment:
> 
> https://github.com/nipy/nibabel/blob/a2e5dee05cf374c22670ff9fd0d385ce366eb495/.github/workflows/test.yml#L199-L212
> 
> I created one here called "Release"

Thanks a lot for the example, I'll move forward in this direction and convert my script to use GitHub Actions.

**Comment by effigies** [comment]
@oesteban Can you make me an admin on the eddymotion PyPI package?

File: .github/workflows/pypi-package.yml [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    steps:
+      - uses: actions/download-artifact@v4
+        with:
+          name: dist
+          path: dist/
+      - uses: pypa/gh-action-pypi-publish@release/v1
+        with:
+          repository-url: https://test.pypi.org/legacy/
+          user: __token__
+          password: ${{ secrets.TESTPYPI_API_TOKEN }}
```
Comment (esavary) [inline .github/workflows/pypi-package.yml:55]: @effigies @oesteban Do we also have an authentication token for TestPyPI? (For now, I've used this generic variable name.)

File: .github/workflows/pypi-package.yml [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    - run: pip install --upgrade build twine
+    - name: Build sdist and wheel
+      run: python -m build
+    - name: Check package
+      run: twine check dist/*
+    - name: Upload sdist and wheel artifacts
+      uses: actions/upload-artifact@v4
+      with:
+        name: dist
+        path: dist/
```
Comment (effigies) [inline .github/workflows/pypi-package.yml:39]:
```suggestion
    name: Build & verify package
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
        with:
          fetch-depth: 0
      - uses: hynek/build-and-inspect-python-package@v2
```

File: .github/workflows/pypi-package.yml [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+  auto-release-test-pypi:
+    runs-on: ubuntu-latest
+    environment: "Test-release"
+    needs: [build-package]
+    if: github.event_name == 'push' && github.ref == 'refs/heads/main'
+    steps:
+      - uses: actions/download-artifact@v4
+        with:
+          name: dist
+          path: dist/
```
Comment (effigies) [inline .github/workflows/pypi-package.yml:50]:
```suggestion
      - name: Download packages built by build-and-inspect-python-package
        uses: actions/download-artifact@v4
        with:
          name: Packages
          path: dist
```

File: .github/workflows/pypi-package.yml (hunk 1 above)
Comment (effigies) [inline .github/workflows/pypi-package.yml:55]: Let's just drop these. I've set up [trusted publishing](https://docs.pypi.org/trusted-publishers/) on test.pypi.org.
```suggestion
```

File: .github/workflows/pypi-package.yml [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+      run: twine check dist/*
+    - name: Upload sdist and wheel artifacts
+      uses: actions/upload-artifact@v4
+      with:
+        name: dist
+        path: dist/
+
+  auto-release-test-pypi:
+    runs-on: ubuntu-latest
+    environment: "Test-release"
```
Comment (effigies) [inline .github/workflows/pypi-package.yml:43]: We don't need an environment for test pushes. It either does nothing or spams maintainers for approval on every push to `main`.

```suggestion
```

File: .github/workflows/pypi-package.yml [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+          repository-url: https://test.pypi.org/legacy/
+          user: __token__
+          password: ${{ secrets.TESTPYPI_API_TOKEN }}
+
+  auto-release-pypi:
+    runs-on: ubuntu-latest
+    environment: "Release"
+    needs: [build]
+    permissions:
+      id-token: write
```
Comment (effigies) [inline .github/workflows/pypi-package.yml:62]: This is already set globally.

```suggestion
```

File: .github/workflows/pypi-package.yml [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    if: github.event_name == 'push' && startsWith(github.ref, 'refs/tags/')
+    steps:
+      - uses: actions/download-artifact@v4
+        with:
+          name: dist
+          path: dist/
```
Comment (effigies) [inline .github/workflows/pypi-package.yml:68]:
```suggestion
      - name: Download packages built by build-and-inspect-python-package
        uses: actions/download-artifact@v4
        with:
          name: Packages
          path: dist
```

File: .github/workflows/pypi-package.yml [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+      - uses: pypa/gh-action-pypi-publish@release/v1
+        with:
+          user: __token__
+          password: ${{ secrets.PYPI_API_TOKEN }}
```
Comment (effigies) [inline .github/workflows/pypi-package.yml:72]: If given admin rights (cc @oesteban) on the PyPI package, I will set up trusted publishing for this, as well.

```suggestion
```

Comment (oesteban) [reply .github/workflows/pypi-package.yml:72]: Done, I believe

File: .github/workflows/pypi-package.yml (hunk 4 above)
Comment (esavary) [reply .github/workflows/pypi-package.yml]: Thanks, indeed, it makes sense. I wasn't sure if we should also restrict the test releases.

**Comment by esavary** [comment]
Is there a clever way to trigger the workflow for testing without pushing changes to the main branch?

File: .github/workflows/pypi-package.yml [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+      - uses: hynek/build-and-inspect-python-package@v2
+
+  auto-release-test-pypi:
+    runs-on: ubuntu-latest
+    needs: [build-package]
+    if: github.event_name == 'push' && github.ref == 'refs/heads/main'
```
Comment (effigies) [inline .github/workflows/pypi-package.yml:30]: > Is there a clever way to trigger the workflow for testing without pushing changes to the main branch?

Comment this out for now:

```suggestion
    # if: github.event_name == 'push' && github.ref == 'refs/heads/main'
```

File: .github/workflows/pypi-package.yml [hunk 9]
```diff
@@ -0,0 +1,52 @@
+---
+name: Build and upload package
+
+on:
+  push:
+    branches: [main]
+    tags: ["*"]
```
Comment (effigies) [inline .github/workflows/pypi-package.yml:7]: Looks like we also need:

```suggestion
  push:
    branches: [main]
    tags: ["*"]
  pull_request:
    branches: [ main ]
```

File: .github/workflows/pypi-package.yml [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+          name: Packages
+          path: dist
+      - uses: pypa/gh-action-pypi-publish@release/v1
+        with:
+          repository-url: https://test.pypi.org/legacy/
+
+  auto-release-pypi:
+    runs-on: ubuntu-latest
+    environment: "Release"
+    needs: [build]
```
Comment (effigies) [inline .github/workflows/pypi-package.yml:46]:
```suggestion
    needs: [build-package]
```

**Comment by effigies** [comment]
Okay, looks like we can't do this from a fork, but it's mostly working.

We will need to adjust our version string, though: test.pypi.org will not accept local identifiers like the `+gcc9a910` in `24.0.0.dev277+gcc9a910`

File: .github/workflows/pypi-package.yml [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+      - uses: hynek/build-and-inspect-python-package@v2
+
+  auto-release-test-pypi:
+    runs-on: ubuntu-latest
+    needs: [build-package]
+    # if: github.event_name == 'push' && github.ref == 'refs/heads/main'
```
Comment (effigies) [inline .github/workflows/pypi-package.yml:32]:
```suggestion
    # if: github.event_name == 'push' && github.ref == 'refs/heads/main'
    if: ${{ github.event_name != 'pull_request' || github.event.pull_request.head.repo.full_name == 'nipreps/eddymotion' }}

```

**Comment by effigies** [comment]
I'm going to merge this. We can fiddle with versions and then resetting the run conditions in a PR from the main repo.

**Review by effigies — APPROVED** [review]

**Pull request merged**